ark/utils/spatial_analysis_utils.py

```diff
@@ -266,6 +257,118 @@ def compute_close_cell_num_random(marker_nums, dist_mat, dist_lim, bootstrap_num
+    # and close_num_random_context we may end up having to do this, then again
+    # this is non-optimized, we can probably look into doing something more sophisticated
+    # in the future
+    num = len(thresh_vec)
+
+    for j in range(num):
+        # we need to regenerate the positive inds per marker so we can compare them
+        # with the positive indices for each cell type, needed so we can bootstrap
+        # in a context-based environment properly
+        marker1posinds = current_fov_channel_data[current_fov_channel_data.columns[j]] > thresh
+
+        # generate the number of positive hits per cell type for a specific marker
+        # in cell_type_facets or else for all the non-cell_type_facets cell types
+        cell_type_nums_per_facet_1 = {}
+        for cell_type, cell_type_data in cell_type_data_per_facet.values():
+            cell_type_nums_per_facet_1[cell_type] = np.sum(
+                np.logical_and(marker1posinds, cell_type_data.index))
+
+        # new iteration, needed to properly generate a pair of 1 vs 2 analysis
+        # basically the same thing as generating cell_type_nums_per_facet_1
+        for k in range(j):
+            marker2posinds = current_fov_channel_data[current_fov_channel_data.columns[k]] > thresh
+
+            cell_type_nums_per_facet_2 = {}
+            for cell_type, cell_type_data in cell_type_data_per_facet.values():
+                cell_type_nums_per_facet_2[cell_type] = np.sum(
+                    np.logical_and(marker2posinds, cell_type_data.index))
+
+                # triple for-loops, what moron put these in...
+                for r in range(bootstrap_num):
+                    # keep track of the random labels we generate for markers 1 and 2
+                    marker1labels_rand = []
+                    marker2labels_rand = []
+
+                    # iterate through each marker type, and do a faceted random sample for 1 and 2
+                    for m, m_data in cell_type_data_per_facet.values():
+                        marker1labelsfacet = np.random.choice(
+                            m_data.values, cell_type_nums_per_facet_1[m]).tolist()
+                        marker1labels_rand.extend(marker1labelsfacet)
+
+                        marker2labelsfacet = np.random.choice(
+                            m_data.values, cell_type_nums_per_facet_2[m]).tolist()
+                        marker2labels_rand.extend(marker2labelsfacet)
+
+                        # now we subset the distance matrix accordingly
+                        # and use dist_lim to threshold
+                        rand_dist_mat = dist_mat.loc[marker1labels_rand, marker2labels_rand].values
+                        rand_dist_mat_bin = np.zeros(rand_dist_mat.shape)
+                        rand_dist_mat_bin[rand_dist_mat < dist_lim] = 1
+
+                        # assign final value to close_num_rand
+                        close_num_rand[j, k, r] = np.sum(rand_dist_mat_bin)
+
+    return close_num_rand
+
```

Comment:
aye, me bones tell me there be some nasty optermizashuns to be made in these choppy loops 🏴‍☠️